### PR TITLE
Prototype quality-based match making

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR proposes a TrueSkill evaluation scheme that
1. queries the match quality against each of the AIs in the existing league,
2. sorts the AIs by their match quality,
3. then picks one of the top three highest-match-quality AI to perform a match.

Empirically, I have found it converge sooner and play more diverse matches.

image

Compared to the binary search approach, this PR's scheme would be more computationally expensive when the league size is large. However, in the situation when we fixed the reference agents, this approach seems to work great empirically.

### CODE REVIEW

**Comment by vwxyzjn** [comment]
@kachayev What are your thoughts on such an evaluation?

**Comment by kachayev** [comment]
@vwxyzjn 

"Quality" here stands for quality estimate done by TrueSkill, right? 

Let me dig a little bit into details of the implementation. I also want to do a quick simulation for doing so in with a small-ish number of agents in the population. The key concern in the matchmaking process is always not to overfit into subpopulations (which is unlikely to happen with such a small number of players). But that needs to be verified. And it's always simple to dodge by having a bernoulli random to allow you to play a random game (the question is on the best parametrization of this). Another concerns from the perspective of selfplay: empirically it's necessary for the agent to play both easy and hard opponents (hard opponents gives you learning signal, easy opponents act as a regularizer against high variance). I need to play around with a system to see i sampling based on quality gives you both. I'll get back as soon as my thoughts are materialized!

File: experiments/quality_league.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+                # sort by quality
+                match_qualities = sorted(match_qualities, key=lambda x: x[1], reverse=True)
+                print("match_qualities[:3]", match_qualities[:3])
+
+                # run a match if the quality of the opponent is high enough
+                top_3_ai = [item[0] for item in match_qualities[:3]]
+                opponent_ai = random.choice(top_3_ai)
+                match_up = (ai.name, opponent_ai.name)
+                match_quality = quality_1vs1(ai, opponent_ai)
+                print(f"the match up is ({ai}, {opponent_ai}), quality is {round(match_quality, 4)}")
```
Comment (vwxyzjn) [inline experiments/quality_league.py:408]: > "Quality" here stands for quality estimate done by TrueSkill, right?

No, so the implementation gets a `quality_score` based on Trueskill for each opponent (via the [`quality_1vs1(r1, r2)` API)](https://trueskill.org/#trueskill.quality_1vs1), sort them by `quality_score`, and randomly samples an opponent from the pool with the 3 highest `quality_score`.

**Comment by vwxyzjn** [comment]
@kachayev I am going to merge this now to run some experiments, but happy to revisit later :D

**Comment by kachayev** [comment]
Note sure this is the most convenient avenue for the discussion (let me know if you want to move this to Discord).

I finally got time to play around with this proposal. I think it's a good idea! Playing games with least predictable outcome should reduce overall uncertainty about skill estimates. Initially I though this is used for matchmaking for training but it seems it's only used for evaluation. Empirically the same approach would be problematic when deciding on opponents for training but just for measurements.

There's still an interesting concern, or even a topic for research/investigations. TrueSkill relies on the prior estimate for draws. For some games, even when following random policy, there is a little chance to get draw (e.g. Chess, or Go). But in other games, it actually takes more sophisticated policy to avoid "drawing" all the time. Specifically keeping in mind practical limitation of using "timestep budget" that typically results in draw. Working with MicroRTS environment we get around this (partially) because we have scripted bots that are quite good from get go. Other options would be an SL policy (imitation learning) to bootstrap the initial population where the estimate of TrueSkill would have reasonably low sigma (and reasonably low prob of drawing). 

Questions now are: 
* how sensitive algorithm to the quality of prior?
* how to properly estimate draws prior for a given population?
* how does it change with training?
* and more

Another question: do we need to re-evaluate skill for existing players when injecting new players into the population. Empirically for large populations, the answer is "yes". It would be interesting to run experiments here for a small size population and only a few newcomers. As we talk about latent variable estimate, we don't ground truth to compare to. But we can run simplified version with only a few match ups selected for each new player vs. full round robin. And measure how different the final result is.

**Pull request merged**